dev/TEAL.md

```diff
@@ -201,6 +201,41 @@ _available_.
 
  * Since v7, the account associated with any contract present in the
    `txn.ForeignApplications` field is _available_.
+   
+ * Since v9, there is group-level resource sharing. Any resource that
+   is available in _some_ top-level transaction in a transaction group
+   is available in _all_ v9 or later application calls in the group,
+   whether those application calls are top-level or inner.
+   
+ * When considering whether an asset holding or application local
+   state is available by group-level resource sharing, the holding or
+   local state must be available in a top-level transaction without
+   considering group sharing. For example, if account A is made
+   available in one transaction, and asset X is made available in
+   another, group resource sharing does _not_ make A's X holding
+   available.
+     
+ * Top-level transactions that are not application calls also make
+   resources available to group-level resource sharing. The following
+   resources are made available other transaction types.
+
+     1. `pay` - `txn.Sender`, `txn.Receiver`, and
+        `txn.CloseRemainderTo` (if set)
+
+     1. `keyreg` - `txn.Sender`
+
+     1. `acfg` - `txn.Sender`, `txn.ConfigAsset`, and the
+        `txn.ConfigAsset` holding of `txn.Sender`.
+
+     1. `axfer` - `txn.Sender`, `txn.AssetReceiver`, `txn.AssetSender`
+        (if set), `txnAssetCloseTo` (if set), `txn.XferAsset` and the
+        `txn.XferAsset` holding of each of those accounts.
```

Comment:
Including the sender account balances, even in case of a clawback?

Reply:
Yes, I decided it was not a big deal to allow a little more access, as this is still much less I/O than an appl could generate. So I decided for simplicity of spec over precisely matching what is actually needed.

(As a further crazy example - axfer and pay of 0 does not actually read in the recipient, but I certainly don't want to spec that.)

Reply:
just curiosity, so you will fetch this information only if an application call needs it? Or always starting v9?

Reply:
The new changes will not cause it to be prefetched.  However, I think we _already_ prefetch this to speed up validation (not proposals, just validating proposals).  That prefetch code does not understand the subtle reasons why the holding might not always be needed. (I don't think.)

At any rate - no change in prefetching will happen due to this change.

Reply:
Actually, that code is fairly clever, so it *won't* be prefetched, before or now.
```
if !stxn.Txn.AssetSender.IsZero() {
	loadAccountsAddResourceTask(nil, basics.CreatableIndex(stxn.Txn.XferAsset), basics.AssetCreatable, task, resourceTasks, queue)
	loadAccountsAddResourceTask(&stxn.Txn.AssetSender, basics.CreatableIndex(stxn.Txn.XferAsset), basics.AssetCreatable, task, resourceTasks, queue)
} else {
	if stxn.Txn.AssetAmount == 0 && (stxn.Txn.AssetReceiver == stxn.Txn.Sender) { // opt in
		loadAccountsAddResourceTask(nil, basics.CreatableIndex(stxn.Txn.XferAsset), basics.AssetCreatable, task, resourceTasks, queue)
	}
	if stxn.Txn.AssetAmount != 0 { // zero transfer is noop
		loadAccountsAddResourceTask(&stxn.Txn.Sender, basics.CreatableIndex(stxn.Txn.XferAsset), basics.AssetCreatable, task, resourceTasks, queue)
	}
}
```